langchain/chains/combine_documents/reduce.py

```diff
@@ -0,0 +1,276 @@
+        **kwargs: Any,
+    ) -> Tuple[str, dict]:
+        """Combine multiple documents recursively.
+
+        Args:
+            docs: List of documents to combine, assumed that each one is less than
+                `token_max`.
+            token_max: Recursively creates groups of documents less than this number
+                of tokens.
+            callbacks: Callbacks to be passed through
+            **kwargs: additional parameters to be passed to LLM calls (like other
+                input variables besides the documents)
+
+        Returns:
+            The first element returned is the single string output. The second
+            element returned is a dictionary of other keys to return.
+        """
+        result_docs, extra_return_dict = self._collapse(
+            docs, token_max, callbacks=callbacks, **kwargs
+        )
+        return self.combine_document_chain.combine_docs(
+            docs=result_docs, callbacks=callbacks, **kwargs
+        )
+
+    async def acombine_docs(
+        self, docs: List[Document], callbacks: Callbacks = None, **kwargs: Any
+    ) -> Tuple[str, dict]:
+        """Combine multiple documents recursively.
+
+        Args:
+            docs: List of documents to combine, assumed that each one is less than
+                `token_max`.
+            token_max: Recursively creates groups of documents less than this number
+                of tokens.
+            callbacks: Callbacks to be passed through
+            **kwargs: additional parameters to be passed to LLM calls (like other
+                input variables besides the documents)
+
+        Returns:
+            The first element returned is the single string output. The second
+            element returned is a dictionary of other keys to return.
+        """
+        result_docs, extra_return_dict = await self._acollapse(
+            docs, callbacks=callbacks, **kwargs
+        )
+        return await self.combine_document_chain.acombine_docs(
+            docs=result_docs, callbacks=callbacks, **kwargs
+        )
+
+    def _collapse(
+        self,
+        docs: List[Document],
+        token_max: int = 3000,
+        callbacks: Callbacks = None,
+        **kwargs: Any,
+    ) -> Tuple[List[Document], dict]:
+        result_docs = docs
+        length_func = self.combine_document_chain.prompt_length
+        num_tokens = length_func(result_docs, **kwargs)
+
+        def _collapse_docs_func(docs: List[Document], **kwargs: Any) -> str:
+            return self._collapse_chain.run(
+                input_documents=docs, callbacks=callbacks, **kwargs
+            )
+
+        while num_tokens is not None and num_tokens > token_max:
+            new_result_doc_list = _split_list_of_docs(
+                result_docs, length_func, token_max, **kwargs
+            )
+            result_docs = []
+            for docs in new_result_doc_list:
+                new_doc = _collapse_docs(docs, _collapse_docs_func, **kwargs)
```

Comment:
why can't we call _collapse_docs_func first and pass output to _collapse_docs (and then we don't need to define _collapse_docs_func in first place)